<?xml version="1.0"?>
<launch>
  <!-- 
    CV Publishers Launch File
    This launch file starts the cv_publishers node with proper topic remapping
  -->

  <!-- Arguments -->
  <arg name="camera_name" default="zed2i" />
  
  <!-- Optional arguments for alternate topic naming -->
  <arg name="use_standardized_topics" default="false" />
  <arg name="rgb_topic" default="/$(arg camera_name)/zed_node/rgb/image_rect_color" />
  <arg name="depth_topic" default="/$(arg camera_name)/zed_node/depth/depth_registered" />
  <arg name="camera_info_topic" default="/$(arg camera_name)/zed_node/rgb/camera_info" />
  <arg name="pose_topic" default="/$(arg camera_name)/zed_node/pose" />
  
  <!-- Topics to use when using standardized names -->
  <arg name="std_rgb_topic" default="/camera/rgb/image_raw" />
  <arg name="std_depth_topic" default="/camera/depth/image_raw" />
  <arg name="std_camera_info_topic" default="/camera/rgb/camera_info" />
  <arg name="std_pose_topic" default="/camera/pose" />

  <!-- Set the appropriate topics based on the use_standardized_topics flag -->
  <arg name="actual_rgb_topic" value="$(eval std_rgb_topic if use_standardized_topics else rgb_topic)" />
  <arg name="actual_depth_topic" value="$(eval std_depth_topic if use_standardized_topics else depth_topic)" />
  <arg name="actual_camera_info_topic" value="$(eval std_camera_info_topic if use_standardized_topics else camera_info_topic)" />
  <arg name="actual_pose_topic" value="$(eval std_pose_topic if use_standardized_topics else pose_topic)" />
  
  <!-- Launch the cv_publishers node with the appropriate topics -->
  <node pkg="autonomy" type="cv_publishers.py" name="cv_publisher" output="screen">
    <param name="rgb_image_topic" value="$(arg actual_rgb_topic)" />
    <param name="depth_image_topic" value="$(arg actual_depth_topic)" />
    <param name="camera_info_topic" value="$(arg actual_camera_info_topic)" />
    <param name="camera_pose_topic" value="$(arg actual_pose_topic)" />
  </node>

  <!-- Launch RViz for visualization (optional) -->
  <arg name="rviz" default="false" />
  <group if="$(arg rviz)">
    <node pkg="rviz" type="rviz" name="rviz" args="-d $(find autonomy)/config/cv_detection.rviz" if="$(eval rviz)">
      <remap from="/tf" to="/tf" />
    </node>
  </group>

</launch>